Change all active TEXT columns in MySQL app DB to LONGTEXT

jeff303 opened the pull request:
Change all active TEXT columns in MySQL app DB to LONGTEXT

Add new text.type Liquibse property to dynamically select the correct text type to use, per DB (which is LONGTEXT for MySQL), very similar to the existing blob.type one

Adding new migrations that update all existing TEXT columns to LONGTEXT columns, for mariadb/mysql only

Update migrations linter to ensure no new text types get added, via a new predicate that searches for text types being added

jeff303 committed:
Change all active TEXT columns in MySQL app DB to LONGTEXT

Add new text.type Liquibase property to dynamically select the correct text type to use, per DB (which is LONGTEXT for MySQL), very similar to the existing blob.type one

Adding new migrations that update all existing app DB TEXT columns to LONGTEXT columns, only for mariadb/mysql

Update migrations linter to ensure no new text types get added, via a new predicate that searches for text types being added

jeff303 committed:
Do the same logic for "blob" now, update rule and test

Revert the previous bulk text type changes

jeff303 committed:
Re-add TEXT migrations after fixing the generation code

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -8648,6 +8657,846 @@ databaseChangeLog:
+  - changeSet:
+      id: 321
+      author: jeff303
+      comment: Added 0.42.0 - modify type of activity.details from text to ${text.type}
+        on mysql,mariadb
```
is this supposed to be part of the comment? Maybe you should make it all one line or use one of the multiline markdown directives just to be safe

jeff303 replied:
This is valid YAML (the "plain style" as described [here](https://stackoverflow.com/a/21699210)).  And this was even generated by our YAML library, using code from here: `https://www.notion.so/metabase/App-DB-Migrations-Tips-and-Tricks-b29b4a8048964ac0a0685b8e6ff19fa0`

But I can see about bulk updating the style to something else instead.

camsaul commented on resources/migrations/000_migrations.yaml:
```diff
@@ -8648,6 +8657,846 @@ databaseChangeLog:
+            newDataType: ${text.type}
+      preConditions:
+        - onFail: MARK_RAN
+        - dbms:
+            type: mysql,mariadb
```
I'm not 100% sure that you can list multiple types here, or if you need to use `or`... have you manually confirmed that this works for both MySQL and MariaDB? Maybe adding a test for MySQL/MariaDB that just confirms the application DB columns are the right types or something could be a good idea

jeff303 replied:
The same thing is used in a couple other places in the file already, [ex](https://github.com/metabase/metabase/blame/a459f19ac8805614acb4709ae16e3aacafb2776a/resources/migrations/000_migrations.yaml#L5342-L5345).  But I can add a test (looks like `test-migrations` macro should work for this).

camsaul replied:
You probably don't even need `test-migrations`, just write a test that checks the type of one of the app DB columns if we're running MySQL or MariaDB. You can use JDBC `DatabaseMetaData` to do that

jeff303 replied:
I just wrote a comprehensive test for all app DBs.

camsaul reviewed (approved):
overall LGTM but I'd try to make all those comments one line to make sure nothing gets confused anywhere

flamber commented:
We still need to create another changeset for 97 (query_cache.results), so existing instances gets converted to LONGBLOB

camsaul commented:
I think we should do the QueryCache migration as a follow-on, mostly because we already have migrations 381 and 382 already so that one would have to be nonsequential anyway

jeff303 committed:
Add missing test for migrations, that checks the new types, for all types in scope for this PR, to ensure that they have all been changed to their expected DB-specific type (either LONGTEXT for MySQL or TEXT/CLOB for others)

Updating a couple migrations to remove special casing that was originally done only for MySQL to simply make them universally ${text.type}, in order to unify behavior and reduce surprises going forward

jeff303 committed:
PR feedback: switch from plain style to block style for comment

jeff303 committed:
Fix namespace ordering

jeff303 commented:
Opened #18815 to cover fixing `query_cache.results` on upgrade.

camsaul committed:
Adopt new migration numbering scheme

camsaul committed:
Adopt new migration numbering scheme

camsaul committed:
Fix comments

camsaul committed:
Update migration numbers

camsaul committed:
Fix MySQL + MariaDB insanity

camsaul committed:
Fix ID range validation

camsaul committed:
Fix ID range validation

camsaul committed:
Actually 382 is the last legacy ID

camsaul committed:
Improved validation and tests

jeff303 committed:
Add `app-db-column-types` helper fn to get col types from app DB

Update ids in test

camsaul committed:
Adopt the new-new migration ID format.

camsaul committed:
Test fixes :wrench:

camsaul committed:
Merge branch 'adopt-new-migration-numbering-scheme' into use-longtext-in-app-db-mysql

camsaul committed:
Merge branch 'master' into use-longtext-in-app-db-mysql

camsaul committed:
Another test fix :wrench:

camsaul commented on enterprise/backend/test/metabase_enterprise/serialization/cmd_test.clj:
```diff
@@ -22,7 +22,7 @@
   [& body]
   `(schema-migrations-test.impl/with-temp-empty-app-db [conn# :h2]
-     (schema-migrations-test.impl/run-migrations-in-range! conn# [0 999999]) ; this should catch all migrations)
+     (schema-migrations-test.impl/run-migrations-in-range! conn# [0 "v99.00-000"]) ; this should catch all migrations)
```
this should have been in the PR to switch to the new migration format but I missed this

camsaul commented on bin/lint-migrations-file/src/change_set/common.clj:
```diff
@@ -15,7 +15,7 @@
 (s/def ::new-style-id
   (s/and string?
-         #(re-matches #"^v\d{2}\.\d{2}-\d{3}$" %)))
+         #(re-matches #"^v\d{2,}\.\d{2}-\d{3}$" %)))
```
this should have been in the PR to switch to the new migration format but I missed this. Minor tweak. Allow >2 digit major versions. So when we get to 100.0 we don't have things start breaking
